packages/hardhat-chai-matchers/src/bigNumber.ts

```diff
@@ -0,0 +1,329 @@
+  Assertion.overwriteMethod("greaterThanOrEqual", gteFunction);
+
+  const lteFunction = override("lte", "be at most", "be above", utils);
+  Assertion.overwriteMethod("most", lteFunction);
+  Assertion.overwriteMethod("lte", lteFunction);
+  Assertion.overwriteMethod("lessThanOrEqual", lteFunction);
+
+  Assertion.overwriteChainableMethod(...createLengthOverride("length"));
+  Assertion.overwriteChainableMethod(...createLengthOverride("lengthOf"));
+
+  Assertion.overwriteMethod("within", overrideWithin(utils));
+
+  Assertion.overwriteMethod("closeTo", overrideCloseTo(utils));
+}
+
+function createLengthOverride(
+  method: string
+): [string, (...args: any[]) => any, (...args: any[]) => any] {
+  return [
+    method,
+    function (_super: any) {
+      return function (this: Chai.AssertionPrototype, value: any) {
+        const actual = this._obj;
+        if (isBigNumber(value)) {
+          const sizeOrLength =
+            actual instanceof Map || actual instanceof Set ? "size" : "length";
+          const actualLength = normalize(actual[sizeOrLength]);
+          const expectedLength = normalize(value);
+          this.assert(
+            actualLength === expectedLength,
+            `expected #{this} to have a ${sizeOrLength} of ${expectedLength.toString()} but got ${actualLength.toString()}`,
+            `expected #{this} not to have a ${sizeOrLength} of ${expectedLength.toString()} but got ${actualLength.toString()}`,
+            actualLength.toString(),
+            expectedLength.toString()
+          );
+        } else {
+          _super.apply(this, arguments);
+        }
+      };
+    },
+    function (_super: any) {
+      return function (this: any) {
+        _super.apply(this, arguments);
+      };
+    } as any,
+  ];
+}
+
+type Methods = "eq" | "gt" | "lt" | "gte" | "lte";
+
+function override(
+  method: Methods,
+  name: string,
+  negativeName: string,
+  utils: Chai.ChaiUtils
+) {
+  return (_super: (...args: any[]) => any) =>
+    overwriteBigNumberFunction(method, name, negativeName, _super, utils);
+}
+
+function normalize(
+  source:
+    | number
+    | bigint
+    | BNType
+    | EthersBigNumberType
+    | BigNumberJsType
+    | string
+): bigint {
+  if (
+    isEthersBigNumber(source) ||
+    isBN(source) ||
+    isBigNumberJsBigNumber(source)
+  ) {
+    return BigInt(source.toString());
+  } else if (
+    typeof source === "string" ||
+    typeof source === "number" ||
+    typeof source === "bigint"
+  ) {
+    // first construct a BigInt, to allow it to throw if there's a fractional value:
+    const toReturn = BigInt(source);
+    // then check the source for integer safety:
+    if (typeof source === "number" && !Number.isSafeInteger(source)) {
+      throw new RangeError(
+        `Cannot compare to unsafe integer ${source}. Consider using BigInt(${source}) instead. For more details, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isSafeInteger`
```

Comment:
I'm pretty sure the suggestion to use `BigInt(${source})` is wrong. We should recommend using `${source}n` instead.

Anyway, this is the function that we'll move to `hardhat/common` and we'll work on a proper explanation and suggestion there.

Reply:
Also: we should tell apart the scenarios where `isSafeInteger` returns false because the number is too large, and the scenarios where it's a non-integer (e.g., `3.14`), and show different error messages in both cases.

Reply:
Oh, I just realized that that's meant to be handled by the `BigInt(source)` call. I think we should be the ones doing the check though, so that we can show a better error.

Reply:
Using the `1n` syntax (rather than the `BigInt(1)` syntax) results in a `tsc` error:
```
test/bigNumber.ts:19:12 - error TS2737: BigInt literals are not available when targeting lower than ES2020.

19     expect(1n).to.equal(new BN(1));
              ~~
```

Reply:
Sorry, I meant in the error message. I think we should do `Consider using ${source}n instead.` instead of `Consider using BigInt(${source}) instead.` Then I mentioned the thing about `3.14` because a message suggesting to use `3.14n` would be really misleading.

Reply:
By the way, I think this is something that we can do more properly when we move this logic to a `hardhat/common` helper, so feel free to disregard this. We can merge this PR and then discuss the error message when we make that change (assuming that you are not going to do that in this PR).

Reply:
Addressed in https://github.com/NomicFoundation/hardhat/pull/2428/commits/c9dafded0e0bebd5fd01395f681c1a186af30cdf